### PR TITLE
merge sv discovery code path, commit 0

### PR DESCRIPTION
**Pull request by SHuang-Broad**
First effort in replacing&mdash;if evaluation of call set from the prototyping tool demonstrates improvements&mdash;current way of emitting SV records with the prototyping tool.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/ImpreciseVariantDetector.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import org.broadinstitute.hellbender.tools.spark.sv.discovery.SvType;
+import org.broadinstitute.hellbender.tools.spark.sv.evidence.EvidenceTargetLink;
+import org.broadinstitute.hellbender.tools.spark.sv.evidence.ReadMetadata;
+import org.broadinstitute.hellbender.tools.spark.sv.utils.PairedStrandedIntervalTree;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ImpreciseVariantDetector {
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/ImpreciseVariantDetector.java]: This logic does more than detect variants, though.. it also annotates existing variants with the imprecise evidence. I'm also a little hesitant to move this all into its own separate class -- we really should be moving towards a model where we look at all three sources of evidence (breakpoint assemblies, imprecise evidence clusters, and coverage) simultaneously for eg @mwalker174 's work, and splitting handling of imprecise evidence into its own class seems like a step in the wrong direction.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java [hunk 2]
```diff
@@ -88,4 +90,100 @@ public void inferSvAndWriteVCF(final String vcfOutputFileName, final String samp
...
+     * Given contig alignments, emit novel adjacency not present on the reference to which the locally-assembled contigs were aligned.
+     */
+    private static Iterator<Tuple2<NovelAdjacencyReferenceLocations, ChimericAlignment>>
+    discoverNovelAdjacencyFromChimericAlignments(final Tuple2<byte[], List<ChimericAlignment>> tigSeqAndChimeras, final SAMSequenceDictionary referenceDictionary) {
+        return Utils.stream(tigSeqAndChimeras._2)
+                .map(ca -> new Tuple2<>(new NovelAdjacencyReferenceLocations(ca, tigSeqAndChimeras._1, referenceDictionary), ca)).iterator();
+    }
+
+
+    // TODO: 1/10/18 to be deprecated
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java]: Can you expand one why and in favor of what in this comment?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java [hunk 3]
```diff
@@ -164,9 +163,71 @@ public void inferSvAndWriteVCF(final JavaRDD<AssemblyContigWithFineTunedAlignmen
...
+                    if (hasNoInsertedSeq) {
+                        type = new SimpleSVType.Deletion(novelAdjacencyReferenceLocations); // clean deletion
+                    } else {
+                        type = new SimpleSVType.Deletion(novelAdjacencyReferenceLocations); // scarred deletion
+                    }
+                } else {
+                    if (hasNoInsertedSeq) {
+                        type = new SimpleSVType.Deletion(novelAdjacencyReferenceLocations); // clean contraction of repeat 2 -> 1, or complex contraction
+                    } else {
+                        throw new GATKException("Something went wrong in type inference, there's suspected deletion happening but both inserted sequence and duplication exits (not supported yet): "
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java]: "exits" -> "exists"? or -> "detected"?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java [hunk 4]
```diff
@@ -164,9 +163,71 @@ public void inferSvAndWriteVCF(final JavaRDD<AssemblyContigWithFineTunedAlignmen
...
+                                + novelAdjacencyReferenceLocations.toString());
+                    }
+                }
+            }
+        } else {
+            type = new SimpleSVType.Inversion(novelAdjacencyReferenceLocations);
+        }
+
+        // developer check to make sure new types are treated correctly
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java]: Can we remove this code block? I don't see how it can actually fail..

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java (hunk 2 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java]: now moved into method `InsDelVariantDetector.produceVariantsFromSimpleChimeras()`, which is called by both `InsDelVariantDetector.inferSvAndWriteVCF()` and `DiscoverVariantsFromContigAlignmentsSAMSpark.discoverVariantsAndWriteVCF()`.

The two calling path now differ by basically how ChimericAlignment is filtered (one of the main difference between existing logic and new logic) and which contigs are sent to `InsDelVariantDetector.produceVariantsFromSimpleChimeras()` (used to be almost all, even contigs with more than two alignments, now only contigs with exactly two alignments).

Exactly how locations and breakpoint complications are inferred for insertions, deletions and small duplications (ones we assembled across the full alt allele) stay the same.

So in short, for simple variants (inversions are not simple any more), which is the vast majority of calls we emit, the two versions will differ only in preprocessing steps, i.e. which contigs are sent for the simple variation inference.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java (hunk 4 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java]: removed

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java (hunk 3 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/InsDelVariantDetector.java]: updated error message.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+                weight = Math.max(alignmentInterval.mapQual/60.0,
+                                  alignmentInterval.alnScore > maxCanonicalChrAlignerScore ? 1 : 0);
+            }
+            tigExplainedQual += weight * len;
+        }
+        return tigExplainedQual;
+    }
+
+    /**
+     * Depending on the number of best-scored configurations, one contig may produce multiple contigs with the same name
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: updated and a todo issued for a bug discovered.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    /**
+     * Manages configuration score-based filtering,
+     * then split the gapped alignments by delegating to {@link ContigAlignmentsModifier#splitGappedAlignment(AlignmentInterval, int, int)}.
+     * Note that this step is essentially a flatMap operation, meaning that one contig may return 1 or multiple contigs:
+     *  *) when 1 contig is yielded, it means the contig has only 1 configuration that scored the best
+     *  *) when multiple contigs are yielded, it means the contig has several top-scored configurations
+     * How to handle the second scenario can be treated in a separate logic unit.
+     */
+    @VisibleForTesting
+    static JavaRDD<AlignedContig> filterAndSplitGappedAI(final JavaRDD<AlignedContig> parsedContigAlignments,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    private static boolean contigNaiveFilter(final AlignedContig contig) {
+        if (contig.alignmentIntervals.size() < 2 ) {
+            return (!contig.alignmentIntervals.isEmpty()) && contig.alignmentIntervals.get(0).mapQual > 20;
+        } else {
+            return contig.alignmentIntervals.stream().mapToInt(ai -> ai.mapQual).filter(mq -> mq > 20).count() > 1;
+        }
+    }
+
+    /**
+     * Manages configuration score-based filtering,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: that was why I called it "manage", but yes it is a bad choice of words; updated.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        toolLogger.info( "Primitive filtering based purely on MQ left " + parsedContigAlignments.count() + " contigs.");
+        return parsedContigAlignments;
+    }
+
+    /**
+     * Idea is to keep mapped contig that has at least two alignments over MQ 20,
+     * or in the case of a single alignment (cannot be simply filtered out the contig because it may contain a large gap),
+     * it must be MQ>20.
+     */
+    private static boolean contigNaiveFilter(final AlignedContig contig) {
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: renamed to `norDiscardForBadMQ`, would this work?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+                        .getAlignedContigs()
+                        .filter(AssemblyContigAlignmentsConfigPicker::contigNaiveFilter).cache();
+        longReads.unpersist();
+        toolLogger.info( "Primitive filtering based purely on MQ left " + parsedContigAlignments.count() + " contigs.");
+        return parsedContigAlignments;
+    }
+
+    /**
+     * Idea is to keep mapped contig that has at least two alignments over MQ 20,
+     * or in the case of a single alignment (cannot be simply filtered out the contig because it may contain a large gap),
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: clarified; and refactored out a static int for the magic number

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        longReads.cache();
+        toolLogger.info( "Processing " + longReads.count() + " raw alignments from " +
+                         longReads.map(GATKRead::getName).distinct().count() + " contigs.");
+
+        final JavaRDD<AlignedContig> parsedContigAlignments =
+                new DiscoverVariantsFromContigAlignmentsSAMSpark.SAMFormattedContigAlignmentParser(longReads, header, false)
+                        .getAlignedContigs()
+                        .filter(AssemblyContigAlignmentsConfigPicker::contigNaiveFilter).cache();
+        longReads.unpersist();
+        toolLogger.info( "Primitive filtering based purely on MQ left " + parsedContigAlignments.count() + " contigs.");
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+                                                                         final Double scoreDiffTolerance,
+                                                                         final Logger toolLogger) {
+
+        final JavaRDD<AlignedContig> parsedContigAlignments = preprocess(longReads, header, toolLogger);
+
+        return filterAndSplitGappedAI(parsedContigAlignments, nonCanonicalContigNamesFile,
+                                      header.getSequenceDictionary(), scoreDiffTolerance);
+    }
+
+    private static JavaRDD<AlignedContig> preprocess(final JavaRDD<GATKRead> longReads, final SAMFileHeader header, final Logger toolLogger) {
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+     * {@link #computeScoreOfConfiguration(List, Set, int)}.
+     *
+     * @param longReads             long read alignments
+     * @param header                header for the long reads
+     * @param scoreDiffTolerance    a tolerance where if two configurations' scores differ less than or equal to this amount, they are considered equally good
+     * @param toolLogger            logger for, most likely, debugging uses
+     *
+     * @return              contigs with alignments filtered and custom formatted as {@link AlignmentInterval}
+     */
+    public static JavaRDD<AlignedContig> parseIntoOptimallyCoveredContig(final JavaRDD<GATKRead> longReads,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: picked the first suggestion, thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+public class AssemblyContigAlignmentsConfigPicker {
+
+    /**
+     * Filters an input of SAM file containing alignments of a single-ended long read that
+     * aims at providing an "optimal coverage" of the long read, based on an heuristic scoring scheme
+     * {@link #computeScoreOfConfiguration(List, Set, int)}.
+     *
+     * @param longReads             long read alignments
+     * @param header                header for the long reads
+     * @param scoreDiffTolerance    a tolerance where if two configurations' scores differ less than or equal to this amount, they are considered equally good
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryDataBundle.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        return broadcastCNVCalls;
+    }
+
+    /**
+     * Loads an external cnv call list and returns the results in an SVIntervalTree. NB: the contig indices in the SVIntervalTree
+     * are based on the sequence indices in the SAM header, _NOT_ the ReadMetadata (which we might not have access to at this
+     * time).
+     */
+    private static SVIntervalTree<VariantContext> loadCNVCalls(final String cnvCallsFile,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryDataBundle.java]: refactored into a new class `CNVInputReader` in package `sv.utils` (in utils because other units could use this functionality as well).

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryDataBundle.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+        this.cnvCallsBroadcast = broadcastCNVCalls(ctx, headerForReads, sampleId, discoverStageArgs);
+        this.assembledIntervals = assembledIntervals;
+        this.evidenceTargetLinks = evidenceTargetLinks;
+        this.metadata = metadata;
+
+        this.toolLogger = toolLogger;
+    }
+
+    private static Broadcast<SVIntervalTree<VariantContext>> broadcastCNVCalls(final JavaSparkContext ctx,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryDataBundle.java]: does the refactoring into the 1st level class `CNVInputReader` address your concern here?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryDataBundle.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+import scala.Tuple2;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.broadinstitute.hellbender.tools.spark.sv.StructuralVariationDiscoveryArgumentCollection.DiscoverVariantsFromContigsAlignmentsSparkArgumentCollection;
+
+
+public final class SvDiscoveryDataBundle {
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryDataBundle.java]: done

**Comment by SHuang-Broad** [comment]
@cwhelan . Thanks for the review!
I've incorporated most of your review suggestions, with the fowling exception because I need to think about what need to be done to make less review rounds

> This logic does more than detect variants, though.. it also annotates existing variants with the imprecise evidence. I'm also a little hesitant to move this all into its own separate class -- we really should be moving towards a model where we look at all three sources of evidence (breakpoint assemblies, imprecise evidence clusters, and coverage) simultaneously for eg @mwalker174 's work, and splitting handling of imprecise evidence into its own class seems like a step in the wrong direction.

I agree. That's what I'm thinking about for complex inversions as well.

So what about the following in this particular PR:

1. move `StructuralVariationDiscoveryPipelineSpark.makeEvidenceLinkTree()` into `ImpreciseVariantDetector`;
2. drop `ImpreciseVariantDetector.detectImpreciseVariantsAndAddReadAnnotations()` considering it really only delegates to `processEvidenceTargetLinks()`
3. rename `ImpreciseVariantDetector` as `EvidenceTargetLinkHandler`
4. reduce the work of `DiscoverVariantsFromContigAlignmentsSAMSpark.discoverVariantsAndWriteVCF()` into detecting only simple variants based on assemblies and name it `discoverSimpleVariants()`
5. let `StructuralVariationDiscoveryPipelineSpark` call into `EvidenceTargetLinkHandler.processEvidenceTargetLinks()` to get back VariantContexts, then write VCF 

`processEvidenceTargetLinks()` really does two things at the moment: annotation on breakpoints and call imprecise deletions; preferably, we should go the all-evidence-at-the-same-time approach and decouple the two but I am trying to not mess with it right now.

If you agree, I'll implement it in a separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/CNVInputReader.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.spark.broadcast.Broadcast;
+import org.broadinstitute.hellbender.engine.FeatureDataSource;
+import org.broadinstitute.hellbender.utils.Utils;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+
+public class CNVInputReader {
+
+    public static Broadcast<SVIntervalTree<VariantContext>> broadcastCNVCalls(final JavaSparkContext ctx,
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/CNVInputReader.java]: I'd still sort of prefer for this method to be on one of the tool classes I think. I think the `Reader` class should just read the calls, validate them, and put them into the interval tree. Then the tool can broadcast them. That way the tool class is the only one that really has to worry about the Spark mechanics of this. Does that make sense to you?

**Comment by cwhelan** [comment]
I think it would be OK to make an `ImpreciseVariantDetector` that is only responsible for returning a list of imprecise variants. But I'd rather not hide the fact that it's annotating the breakpoint-detected variants inside that class. Why not break up these two things and push the annotation close to where the breakpoint-detected variants are created (ie. in `discoverSimpleVariants` or something), and then call the imprecise variant detector after that?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/CNVInputReader.java (hunk 17 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/CNVInputReader.java]: Makes sense. 
Just to make sure I understand the request: it is OK to have a 
`Broadcast<SVIntervalTree<VariantContext>> cnvCallsBroadcast` in `SvDiscoveryInputData`, just to move the broadcasting method into an actual tool. Am I understanding correctly?

Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/CNVInputReader.java]: now refactored

**Comment by SHuang-Broad** [comment]
ticked opened in #4189
